kvserver: create LinkExternalSSTRequest kv rpc

msbutler opened the pull request:
Previously, the AddSStable rpc would process linking external files into
pebble, even though the linking logic is largely seperate from a regular
AddSSTable request. This patch splits external file linking into a new rpc:
LinkExternalSST.

After this patch, the only significant code overlap between the AddSStable and
LinkExternalSST rpcs is in raft sideloading.

Fixes https://github.com/cockroachdb/cockroach/issues/120526

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/120570)
<!-- Reviewable:end -->

msbutler commented on pkg/kv/kvpb/api.proto:
```diff
@@ -2200,6 +2165,47 @@ message AddSSTableResponse {
+  // Including these stats can make the evaluation of LinkExternalSSTable much cheaper.
+  storage.enginepb.MVCCStats mvcc_stats = 3 [(gogoproto.customname) = "MVCCStats"];
+
+  message  ExternalFile {
```
Thoughts on  either:
- flattening this proto (i.e. get rid of the `ExternalFile` message)? 
- or, we could use the `ExternalFile` message to only describe the backing file (i.e. we'd remove the `SyntheticPrefix` field) and use this message in the `proposer_kv.proto`.

dt replied:
yeah, I'd flatten it.

dt replied:
well, wait, no, we likely want to be able to repeat external files when we restore inc backups. I'd keep it as is.

msbutler commented on pkg/kv/kvserver/replica_application_cmd.go:
```diff
@@ -100,6 +100,7 @@ func (c *replicatedCmd) getStoreWriteByteSizes() (writeBytes int64, ingestedByte
 		writeBytes = int64(len(c.Cmd.WriteBatch.Data))
 	}
+	// TODO(msbutler): do we need extra logic for external ssts?
 	if c.Cmd.ReplicatedEvalResult.AddSSTable != nil {
```
I believe the answer here is no: `writeBytes` and `ingestedBytes` should be 0 for an external file.

msbutler replied:
Going with 0, to keep the accounting consistent with AddSStable: we only increment for data, not metadata.

msbutler commented on pkg/kv/kvserver/replica_raft.go:
```diff
@@ -181,6 +181,7 @@ func (r *Replica) evalAndPropose(
 	// granularity of admission control doing token and size estimation (which
 	// is 15s). Also, admission control corrects for gaps in reporting.
 	writeBytes := kvadmission.NewStoreWriteBytes()
+	// TODO(msbutler): anything to do here for external sst linking?
```
i believe its fine to tell admission control that the writeBytes and ingestedBytes is 0.

dt replied:
yes.

msbutler commented on pkg/kv/kvserver/kvserverpb/proposer_kv.proto:
```diff
@@ -259,6 +254,18 @@ message ReplicatedEvalResult {
+    roachpb.Span span = 1 [(gogoproto.nullable) = false];
+    string remote_file_loc = 2;
+    string remote_file_path = 3;
+    uint64 backing_file_size = 4;
```
we no longer use BackingFileSize. can we stop plumbing this?

dt replied:
I think we want it; it can avoid a stat call during open if we have it.

msbutler replied:
I don't think we currently use it though. Are you saying we'd use it here? https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/backupccl/restore_online.go#L300

dt replied:
We _set_ it here: https://github.com/cockroachdb/cockroach/blob/6681efbbd7d49f09167febc9ef8db01b191a77ab/pkg/ccl/backupccl/restore_online.go#L296

I'm saying we can use it in pebble, when we open the SST, to avoid needing to stat it to know how far back to open at to find the footer.

dt replied:
I'd just leave it for now, and keep this PR a pure code-move to split the RPC and raft proposal paths out of AddSSTable with no behavior change.

msbutler commented:
unrelated flakes

msbutler commented on pkg/kv/kvserver/kvserverpb/proposer_kv.proto:
```diff
@@ -258,6 +252,22 @@ message ReplicatedEvalResult {
+  // LinkExternalSSTable, like AddSSTable, is a side effect that must execute
+  // before the Raft application is committed. It must be idempotent to account
+  // for an ill-timed crash after applying the side effect, but before
+  // committing the batch.
```
to be transparent: we considered reducing the scope of this pr to be only above raft: i.e. cmd_link_external_sstable would create a addsstable replicated eval result. One reason we want to plumb the logic below raft is because it makes it more clear that regular addsstable requests _always_ have data to sideload.

dt commented on pkg/kv/kvpb/api.proto:
```diff
@@ -2200,6 +2165,53 @@ message AddSSTableResponse {
+  // MVCCStats are estimated MVCCStats for the contents of this Extrenal SSTable
+  // and is used as-is during evaluation of the command to update the range
+  // MVCCStats, instead of computing the stats for the SSTable by iterating it.
+  storage.enginepb.MVCCStats mvcc_stats = 3 [(gogoproto.customname) = "MVCCStats"];
```
Maybe move this into ExternalFile so if/when we repeat it, the stats are per-file, not per-request.

dt commented on pkg/kv/kvpb/api.proto:
```diff
@@ -2200,6 +2165,53 @@ message AddSSTableResponse {
+    uint64 approximate_physical_size = 4;
+    bytes synthetic_prefix = 5;
+  }
+  ExternalFile external_file = 4 [(gogoproto.nullable) = false];
```
I'd make this `repeated` then just make it an error response if it is len()!=1 for now, but that way it is clearer that folks should add new file-specific fields in the file struct.

msbutler replied:
hrm, is this pattern used elsewhere?

dt commented on pkg/kv/kvserver/kvserverpb/proposer_kv.proto:
```diff
@@ -258,6 +252,22 @@ message ReplicatedEvalResult {
+  // before the Raft application is committed. It must be idempotent to account
+  // for an ill-timed crash after applying the side effect, but before
+  // committing the batch.
+  message LinkExternalSSTable {
```
this struct does differ from the request one, e.g. mvcc stats need to make it to kv server but don't go into (this part) of the proposal so I think this justifies the somewhat duplicated structs.

msbutler replied:
i.e. we shouldn't pass the `ExternalFile` message currently defined in api.proto? agreed.

dt commented on pkg/kv/kvserver/replica_app_batch.go:
```diff
@@ -288,6 +288,13 @@ func (b *replicaAppBatch) runPostAddTriggersReplicaOnly(
 		res.AddSSTable = nil
 	}
+	if res.LinkExternalSSTable != nil {
+		if res.LinkExternalSSTable.RemoteRewriteTimestamp.IsSet() {
```
What's this timestamp check for?

msbutler replied:
i was copying the AddSSTable req code above. I'll copy the comment: 
```
// We've ingested the SST already (via the appBatch), so all that's left
// to do here is notify the rangefeed, if appropriate.
```

msbutler replied:
And we check the timestamp bc non mvcc ExternalSSTable requests are not sent to rangefeeds.

nvb reviewed (commented):
> Reviewed 17 of 21 files at r1, 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120570)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @erikgrinaker, and @msbutler)

___
*[`pkg/kv/kvpb/method.go` line 142 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXs8aTBtRKUl_85N1U:-NtXs8aTBtRKUl_85N1V:b6gvz8x) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvpb/method.go#L142)):*
> ```Go
> 	// Best-effort.
> 	AdminScatter
> 	// AddSSTable links a file into the RocksDB log-structured merge-tree.
> ```

What's RocksDB? The only log-structured merge tree I know is pebble.
___
*[`pkg/kv/kvserver/app_batch.go` line 186 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXsP5l4tabVbRnupXC:-NtXsP5l4tabVbRnupXD:b-3rd5hy) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvserver/app_batch.go#L186)):*
> ```Go
> 		)
> 		b.updateSSTStats(res, copied)
> 	} else if res.LinkExternalSSTable != nil {
> ```

nit: I don't think this needs to be an `else`, it can just be a standalone `if`. That will keep this consistent with the rest of post-application side effect handling and avoid an unnecessary assumption about mutual exclusive handling of these fields.
___
*[`pkg/kv/kvserver/replica_app_batch.go` line 293 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXt9gx3STHJGbtg5Ea:-NtXt9gx3STHJGbtg5Eb:b-m4bfhe) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvserver/replica_app_batch.go#L293)):*
> ```Go
> 	if res.LinkExternalSSTable != nil {
> 		if res.LinkExternalSSTable.RemoteRewriteTimestamp.IsSet() {
> 			b.r.handleSSTableRaftMuLocked(ctx, nil, res.LinkExternalSSTable.Span, res.WriteTimestamp)
> ```

This is going to publish the contents of the `sst` onto any rangefeeds. Is that the desired effect?
___
*[`pkg/kv/kvserver/replica_app_batch.go` line 296 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXt3zN5w8RwdHEMYQi:-NtXt3zN5w8RwdHEMYQj:b6i8m5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvserver/replica_app_batch.go#L296)):*
> ```Go
> 		}
> 		res.LinkExternalSSTable = nil
> ```

nit: stray line
___
*[`pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go` line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXtShXBMao_MN99Fxm:-NtXtShXBMao_MN99Fxn:bb6ht39) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go#L1)):*
> ```Go
> // Copyright 2017 The Cockroach Authors.
> ```

2024
___
*[`pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go` line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXteIuBPwOt3N5JT1d:-NtXteIuBPwOt3N5JT1e:b-siesz5) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go#L33)):*
> ```Go
> }
> 
> // EvalLinkExternalSSTable evaluates an LinkExternalSSTable command. For details, see doc comment
> ```

s/an/a/
___
*[`pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go` line 44 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXtoxbBGsHI2Cum2me:-NtXtoxbBGsHI2Cum2mf:b-hqdtkl) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go#L44)):*
> ```Go
> 
> 	var span *tracing.Span
> 	ctx, span = tracing.ChildSpan(ctx, "LinkExternalSSTable")
> ```

Do we need this tracing span? The evaluation looks pretty cheap.
___
*[`pkg/kv/kvserver/raftlog/payload.go` line 48 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXv6KdDMPmxXUStIIa:-NtXv6KdDMPmxXUStIIb:b-d6s7ij) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvserver/raftlog/payload.go#L48)):*
> ```Go
> 		// raft.ConfChange.
> 		prefix = false
> 	} else if command.ReplicatedEvalResult.AddSSTable != nil || command.ReplicatedEvalResult.LinkExternalSSTable != nil {
> ```

@dt and I discussed this. There's not a good reason to sideload `LinkExternalSSTable` entries. Doing so leads to a 0 byte file being written and read and requires all of the new `addSSTable != nil` checks. I think we can skip sideloading without much work and avoid all of this — just revert the changes this PR is making to `payload.go` and `sideload.go`.


<!-- Sent from Reviewable.io -->

msbutler committed:
kvserver: create LinkExternalSSTRequest rpc

Previously, the AddSStable rpc would process linking external files into
pebble, even though the linking logic is largely seperate from a regular
AddSSTable request. This patch splits external file linking into a new rpc:
LinkExternalSST.

NB: after this patch, External SSTs will no longer sideload into each pebble
instance.

Fixes #120526

Release note: none

msbutler reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120570)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @erikgrinaker, and @nvanbenschoten)

___
*[`pkg/kv/kvpb/method.go` line 142 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXs8aTBtRKUl_85N1U:-NtaRiba7CcMrDeCIUpI:blrwwyz) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvpb/method.go#L142)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What's RocksDB? The only log-structured merge tree I know is pebble.
</blockquote></details>

Ah good q. RocksDB was crdb's old storage engine. The blog post below explains why we built Pebble to replace RocksDB. I'll update the comment to prevent future confusion :)

https://www.cockroachlabs.com/blog/pebble-rocksdb-kv-store/
___
*[`pkg/kv/kvserver/app_batch.go` line 186 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXsP5l4tabVbRnupXC:-NtaU30C8wInnLwFrtWf:b-xz242v) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvserver/app_batch.go#L186)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: I don't think this needs to be an `else`, it can just be a standalone `if`. That will keep this consistent with the rest of post-application side effect handling and avoid an unnecessary assumption about mutual exclusive handling of these fields.
</blockquote></details>

Done.

One follow up q: do we still need to update sst stats here if we no longer sideload External SSTs? The comment above may be referring to sideloading, but I don't know enough about the mechanical differences between sideloading and ingesting an sst into pebble. 
___
*[`pkg/kv/kvserver/replica_app_batch.go` line 293 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXt9gx3STHJGbtg5Ea:-NtaUNVFEYjATG0him3i:b-i5r7nt) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvserver/replica_app_batch.go#L293)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is going to publish the contents of the `sst` onto any rangefeeds. Is that the desired effect?
</blockquote></details>

yeah i think so. If the online restore is running from a source tenant in a physical cluster replication setup, these SSTs must be replicated to the standby tenant. 
___
*[`pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go` line 44 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXtoxbBGsHI2Cum2me:-NtaZ2zSB_aGNcW7pssp:b-86rbbr) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go#L44)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we need this tracing span? The evaluation looks pretty cheap.
</blockquote></details>

Good point. Removing.
___
*[`pkg/kv/kvserver/kvserverpb/proposer_kv.proto` line 261 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120570#gh-1526805541:-NtaZgvQEdrS9gcvzvCT:bp7vbga) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae447ea197ebea55374b118a7fd8274a161fef3a/pkg/kv/kvserver/kvserverpb/proposer_kv.proto#L261)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I'd just leave it for now, and keep this PR a pure code-move to split the RPC and raft proposal paths out of AddSSTable with no behavior change.
</blockquote></details>

sounds good.
___
*[`pkg/kv/kvserver/raftlog/payload.go` line 48 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXv6KdDMPmxXUStIIa:-NtaZjFd72PmXW9U2rwJ:b-bqk2nu) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvserver/raftlog/payload.go#L48)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

@dt and I discussed this. There's not a good reason to sideload `LinkExternalSSTable` entries. Doing so leads to a 0 byte file being written and read and requires all of the new `addSSTable != nil` checks. I think we can skip sideloading without much work and avoid all of this — just revert the changes this PR is making to `payload.go` and `sideload.go`.
</blockquote></details>

Will do. 
___
*[`pkg/kv/kvserver/replica_raft.go` line 184 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120570#gh-1526800628:-NtaX01r4DfP1zZAbNmg:b-dg0bhi) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae447ea197ebea55374b118a7fd8274a161fef3a/pkg/kv/kvserver/replica_raft.go#L184)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

yes, or perhaps the size of the struct and thus the underlying manifest write? 
</blockquote></details>

Going with 0, to keep the accounting consistent with AddSStable: we only increment for data, not metadata. @nvanbenschoten if you disagree, let us know. 


<!-- Sent from Reviewable.io -->

msbutler commented:
rebased on master

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 10 of 15 files at r3, 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-:-Ntb1FRN8qZhYIRRip0v:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dt, @erikgrinaker, and @msbutler)

___
*[`pkg/kv/kvpb/method.go` line 142 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXs8aTBtRKUl_85N1U:-Ntb1Jcc4zFLaVt04bm9:b-ffnp8w) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvpb/method.go#L142)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

Ah good q. RocksDB was crdb's old storage engine. The blog post below explains why we built Pebble to replace RocksDB. I'll update the comment to prevent future confusion :)

https://www.cockroachlabs.com/blog/pebble-rocksdb-kv-store/
</blockquote></details>

One day we'll finish getting rid of all of these stale references :)
___
*[`pkg/kv/kvserver/app_batch.go` line 186 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXsP5l4tabVbRnupXC:-Ntb-6DP0JS6yDjhp1ce:bnsj1vf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvserver/app_batch.go#L186)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

Done.

One follow up q: do we still need to update sst stats here if we no longer sideload External SSTs? The comment above may be referring to sideloading, but I don't know enough about the mechanical differences between sideloading and ingesting an sst into pebble.
</blockquote></details>

I think you probably want to handle stats differently for `LinkExternalSSTable`. `numAddSST` and `numAddSSTCopies` are used to drive updates the `AddSSTableApplications` and `AddSSTableApplicationCopies` counters. `numMutations` is used to update load stats for load-based rebalancing. I don't think we want to update the first two, and based on a conversation with @dt, not the third either. So we can probably start by inlining `updateSSTStats` in the `AddSSTable` branch, getting rid of the helper, and not updating these fields on the `LinkExternalSSTable` path.
___
*[`pkg/kv/kvserver/replica_app_batch.go` line 293 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXt9gx3STHJGbtg5Ea:-Ntb0IyyC-Jm6jJEDbVj:b-4n1ui5) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvserver/replica_app_batch.go#L293)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

yeah i think so. If the online restore is running from a source tenant in a physical cluster replication setup, these SSTs must be replicated to the standby tenant.
</blockquote></details>

Will this work though? We're passing `nil` for `sst []byte`. It doesn't looks like `ConsumeSSTable` is handling this case, and if it does, does that mean it's going to download the external file synchronously in order to publish it on a rangefeed?
___
*[`pkg/kv/kvserver/replica_raft.go` line 184 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120570#gh-1526800628:-Ntb1C0tELbv0MU9QwbT:b-otgo0k) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae447ea197ebea55374b118a7fd8274a161fef3a/pkg/kv/kvserver/replica_raft.go#L184)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

Going with 0, to keep the accounting consistent with AddSStable: we only increment for data, not metadata. @nvanbenschoten if you disagree, let us know.
</blockquote></details>

0 SGTM.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> Reviewed 5 of 21 files at r1, 1 of 5 files at r2, 10 of 15 files at r3, 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120570)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dt, @msbutler, and @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_app_batch.go` line 293 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXt9gx3STHJGbtg5Ea:-NtbAhHxDu04_sjxnBC3:bbakmri) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvserver/replica_app_batch.go#L293)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Will this work though? We're passing `nil` for `sst []byte`. It doesn't looks like `ConsumeSSTable` is handling this case, and if it does, does that mean it's going to download the external file synchronously in order to publish it on a rangefeed?
</blockquote></details>

What @nvanbenschoten said.
___
*[`pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go` line 48 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-Ntb9kUt7gWoETmzqp_G:-Ntb9kUt7gWoETmzqp_H:b-s83vsi) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7085129d1c0f006246213d8b1d44995530a42b2/pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go#L48)):*
> ```Go
> 
> 	// We have no idea if the SST being ingested contains keys that will shadow
> 	// existing keys or not, so we need to force its mvcc stats to be estimates.
> ```

Is the common case an empty span? Is it worth probing the keyspace to see if it's empty first, and only mark as estimate if it isn't empty?
___
*[`pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go` line 66 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-Ntb9yy43f7Z4zxmUfkJ:-Ntb9yy43f7Z4zxmUfkK:b-5v02ev) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7085129d1c0f006246213d8b1d44995530a42b2/pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go#L66)):*
> ```Go
> 			// Since the remote SST could contain keys at any timestamp, consider it
> 			// a history mutation.
> 			MVCCHistoryMutation: &kvserverpb.ReplicatedEvalResult_MVCCHistoryMutation{
> ```

Is this still a history mutation when `RemoteRewriteTimestamp` is used? Is it because we could still write below other keys?

This will cause a rangefeed to error when ingested. Is that the desired result? I guess it has to be if the above is true.


<!-- Sent from Reviewable.io -->

msbutler reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120570)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dt, @erikgrinaker, and @nvanbenschoten)

___
*[`pkg/kv/kvserver/app_batch.go` line 186 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXsP5l4tabVbRnupXC:-Ntb8HSgBDUkGXnFaeQm:b-hsgwky) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvserver/app_batch.go#L186)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I think you probably want to handle stats differently for `LinkExternalSSTable`. `numAddSST` and `numAddSSTCopies` are used to drive updates the `AddSSTableApplications` and `AddSSTableApplicationCopies` counters. `numMutations` is used to update load stats for load-based rebalancing. I don't think we want to update the first two, and based on a conversation with @dt, not the third either. So we can probably start by inlining `updateSSTStats` in the `AddSSTable` branch, getting rid of the helper, and not updating these fields on the `LinkExternalSSTable` path.
</blockquote></details>

Thanks for the clarification. Done.
___
*[`pkg/kv/kvserver/replica_app_batch.go` line 293 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXt9gx3STHJGbtg5Ea:-Ntb9JM0BwADUZYhUi_8:b-rtxfz4) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvserver/replica_app_batch.go#L293)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

What @nvanbenschoten said.
</blockquote></details>

hrm. It does seem that C2C would currently panic. I'll remove this for now. I'll continue to nil out `res.LinkExternalSSTable` here to follow the `AddSStable` pattern. 
___
*[`pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go` line 48 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-Ntb9kUt7gWoETmzqp_G:-NtbEJTg4UH9qfwQI7i8:bn498l8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7085129d1c0f006246213d8b1d44995530a42b2/pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go#L48)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Is the common case an empty span? Is it worth probing the keyspace to see if it's empty first, and only mark as estimate if it isn't empty?
</blockquote></details>

Ah, this comment should be updated. MVCCStats are always going to be estimates, regardless of the empty key space question, as they're currently hacked together from backup file stats here.
https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/backupccl/restore_online.go#L303
___
*[`pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go` line 66 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-Ntb9yy43f7Z4zxmUfkJ:-NtbC0Bj7J_sp0mxRIjb:b-1vpky9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7085129d1c0f006246213d8b1d44995530a42b2/pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go#L66)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Is this still a history mutation when `RemoteRewriteTimestamp` is used? Is it because we could still write below other keys?

This will cause a rangefeed to error when ingested. Is that the desired result? I guess it has to be if the above is true.
</blockquote></details>

Oof, really nice catch. It is no longer a history mutation if `RemoteRewriteTimestamp` is used. 


<!-- Sent from Reviewable.io -->

msbutler reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120570)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt, @erikgrinaker, and @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_app_batch.go` line 293 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXt9gx3STHJGbtg5Ea:-NtbKuLh2n7qGnVkNqZQ:b-cv1sjl) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvserver/replica_app_batch.go#L293)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

hrm. It does seem that C2C would currently panic. I'll remove this for now. I'll continue to nil out `res.LinkExternalSSTable` here to follow the `AddSStable` pattern.
</blockquote></details>

one follow up thought: if ingested external ssts cannot not be pushed onto rangefeeds for now, i feel like we want the rangefeed to error out instead of silently ignoring these linked files. To get this behavior, should we universally set `MVCCHistoryMutation` during eval? cc @dt 


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> Reviewed 1 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120570)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt, @msbutler, and @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_app_batch.go` line 293 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXt9gx3STHJGbtg5Ea:-NtbISWT3QBmtRzy9om6:b-4z3nnl) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvserver/replica_app_batch.go#L293)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

hrm. It does seem that C2C would currently panic. I'll remove this for now. I'll continue to nil out `res.LinkExternalSSTable` here to follow the `AddSStable` pattern.
</blockquote></details>

Consider adding a comment for why we don't do anything here.
___
*[`pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go` line 61 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtbLbIvEGvTubx6ttve:-NtbLbIvEGvTubx6ttvf:b-90toab) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7085129d1c0f006246213d8b1d44995530a42b2/pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go#L61)):*
> ```Go
> 				BackingFileSize:         args.ExternalFile.BackingFileSize,
> 				Span:                    roachpb.Span{Key: start.Key, EndKey: end.Key},
> 				RemoteRewriteTimestamp:  sstToReqTS,
> ```

I'm not sure this rewrite works correctly.

With AddSSTable, this gives the timestamp written in the SST. If this timestamp is different from the request timestamp (e.g. because the request got pushed), the SST must be rewritten at the _request_ timestamp, otherwise it can e.g. violate the closed timestamp. I don't see us handling that case here anywhere and plumbing through the request timestamp instead.
___
*[`pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go` line 66 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-Ntb9yy43f7Z4zxmUfkJ:-NtbJb2Y3fhz3iGYeKEk:b-gptkgr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7085129d1c0f006246213d8b1d44995530a42b2/pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go#L66)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

Oof, really nice catch. It is no longer a history mutation if `RemoteRewriteTimestamp` is used.
</blockquote></details>

Ok. We don't check for conflicts here though, so we can write below an existing value -- is that right and ok?

I think this still enforces that we write above the closed timestamp, so it should still work with rangefeeds and such.


<!-- Sent from Reviewable.io -->

msbutler reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120570)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt, @erikgrinaker, and @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_app_batch.go` line 293 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXt9gx3STHJGbtg5Ea:-NtbRutD1UyyPzckl_2T:b-t9scxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvserver/replica_app_batch.go#L293)):*
Done. Wdyt of my previous q here? 

>if ingested external ssts cannot not be pushed onto rangefeeds for now, i feel like we want the rangefeed to error out instead of silently ignoring these linked files. To get this behavior, should we universally set `MVCCHistoryMutation` during eval?
___
*[`pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go` line 61 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtbLbIvEGvTubx6ttve:-NtbQ-B85g2GF2QG50Oj:b60828v) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7085129d1c0f006246213d8b1d44995530a42b2/pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go#L61)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

I'm not sure this rewrite works correctly.

With AddSSTable, this gives the timestamp written in the SST. If this timestamp is different from the request timestamp (e.g. because the request got pushed), the SST must be rewritten at the _request_ timestamp, otherwise it can e.g. violate the closed timestamp. I don't see us handling that case here anywhere and plumbing through the request timestamp instead.
</blockquote></details>

Ah. Another nice catch. Fixed.
___
*[`pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go` line 66 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-Ntb9yy43f7Z4zxmUfkJ:-NtbQ5b05bR1KvVKlcsE:bax323u) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7085129d1c0f006246213d8b1d44995530a42b2/pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go#L66)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Ok. We don't check for conflicts here though, so we can write below an existing value -- is that right and ok?

I think this still enforces that we write above the closed timestamp, so it should still work with rangefeeds and such.
</blockquote></details>

I don't think this new request needs to check for conflicts -- addsstable requests in conventional restore don't check for conflicts (at least in release builds), and further the client should never write into existing key space below an existing value.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (approved):
> Reviewed 1 of 4 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120570)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt, @msbutler, and @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_app_batch.go` line 293 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXt9gx3STHJGbtg5Ea:-NtbXONF7K3FsCCazFCK:b9an4u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvserver/replica_app_batch.go#L293)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

Done. Wdyt of my previous q here? 

>if ingested external ssts cannot not be pushed onto rangefeeds for now, i feel like we want the rangefeed to error out instead of silently ignoring these linked files. To get this behavior, should we universally set `MVCCHistoryMutation` during eval?
</blockquote></details>

I'd rather we returned an appropriate explicit error in this case. Something like:

```go
if res.LinkExternalSSTable != nil {
	b.r.disconnectRangefeedSpanWithErr(res.LinkExternalSSTable.Span, errors.New("LinkExternalSSTable not supported in rangefeeds"))
	res.LinkExternalSSTable = nil
}
```
___
*[`pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go` line 61 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtbLbIvEGvTubx6ttve:-NtbYTDS1LlA0M8xPYDc:b5se7w1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7085129d1c0f006246213d8b1d44995530a42b2/pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go#L61)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

Ah. Another nice catch. Fixed.
</blockquote></details>

nit: replacing `sstToReqTS` seems a bit iffy here, and we don't actually need it either since we don't appear to care about the original timestamp in the SSTs (this was a safety mechanism in AddSSTable to make sure we didn't clobber SSTs unintentionally). Something like:

```go
var rewriteTimestamp hlc.Timestamp
if sstToReqTS.IsSet() {
  // LinkExternalSSTable doesn't care about the original SST timestamp, so just
  // always use the request write timestamp.
  rewriteTimestamp = cArgs.Header.Timestamp 
}
```
___
*[`pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go` line 66 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-Ntb9yy43f7Z4zxmUfkJ:-NtbZ8dN-3ZtM4D5sIkC:b-4hvi0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7085129d1c0f006246213d8b1d44995530a42b2/pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go#L66)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

I don't think this new request needs to check for conflicts -- addsstable requests in conventional restore don't check for conflicts (at least in release builds), and further the client should never write into existing key space below an existing value.
</blockquote></details>

Got it.


<!-- Sent from Reviewable.io -->

msbutler reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120570)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt, @erikgrinaker, and @nvanbenschoten)

___
*[`pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go` line 61 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtbLbIvEGvTubx6ttve:-Ntbf4StCN-8gFyUWT5d:b-a12p2x) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7085129d1c0f006246213d8b1d44995530a42b2/pkg/kv/kvserver/batcheval/cmd_link_external_sstable.go#L61)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

nit: replacing `sstToReqTS` seems a bit iffy here, and we don't actually need it either since we don't appear to care about the original timestamp in the SSTs (this was a safety mechanism in AddSSTable to make sure we didn't clobber SSTs unintentionally). Something like:

```go
var rewriteTimestamp hlc.Timestamp
if sstToReqTS.IsSet() {
  // LinkExternalSSTable doesn't care about the original SST timestamp, so just
  // always use the request write timestamp.
  rewriteTimestamp = cArgs.Header.Timestamp 
}
```
</blockquote></details>

Nice idea. that cleans up the proto: no need to manually pass a timestamp anymore!


<!-- Sent from Reviewable.io -->

msbutler reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120570)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt, @erikgrinaker, and @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_app_batch.go` line 293 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120570#-NtXt9gx3STHJGbtg5Ea:-NtbfL4B4ekMFVpXHZ7u:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b91fd14df9291d34fdbe5268c184abb99d4a147/pkg/kv/kvserver/replica_app_batch.go#L293)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

I'd rather we returned an appropriate explicit error in this case. Something like:

```go
if res.LinkExternalSSTable != nil {
	b.r.disconnectRangefeedSpanWithErr(res.LinkExternalSSTable.Span, errors.New("LinkExternalSSTable not supported in rangefeeds"))
	res.LinkExternalSSTable = nil
}
```
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

msbutler commented:
TFTRs!!

bors r+
